pkg/apis/deviceplugin/v1/gpudeviceplugin_webhook.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2020 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/version"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var (
+	// log is for logging in this package.
+	gpudevicepluginlog = logf.Log.WithName("gpudeviceplugin-resource")
+
+	gpuMinVersion = version.MustParseSemantic("0.18.0")
+)
+
+// SetupWebhookWithManager sets up a webhook for GpuDevicePlugin custom resources.
+func (r *GpuDevicePlugin) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate-deviceplugin-intel-com-v1-gpudeviceplugin,mutating=true,failurePolicy=fail,groups=deviceplugin.intel.com,resources=gpudeviceplugins,verbs=create;update,versions=v1,name=mgpudeviceplugin.kb.io
+
+var _ webhook.Defaulter = &GpuDevicePlugin{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *GpuDevicePlugin) Default() {
+	gpudevicepluginlog.Info("default", "name", r.Name)
+
+	if len(r.Spec.Image) == 0 {
+		r.Spec.Image = "intel/intel-gpu-plugin:0.18.0"
+	}
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-deviceplugin-intel-com-v1-gpudeviceplugin,mutating=false,failurePolicy=fail,groups=deviceplugin.intel.com,resources=gpudeviceplugins,versions=v1,name=vgpudeviceplugin.kb.io
+
+var _ webhook.Validator = &GpuDevicePlugin{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *GpuDevicePlugin) ValidateCreate() error {
```

Comment:
can we validate if a CR already exists with the `NodeSelector` settings because the deployments are going to fail (a device plugin already running on that node)

Reply:
Could you please elaborate? You think the DaemonSet controller can submit a pod to the same node twice?

Reply:
I deployed two `QatDevicePlugin`s. The latter created another daemonset and the pods for it fail for obvious reasons. My comment was can we try to validate new CRs if we see they're going to land pods that will fail

Reply:
Right, the current validator is too permissive.

It's possible to get list of nodes and to calculate intersections. I'm thinking how to do it best, because kubebuilder's webhooks are stateless.

Reply:
@mythi What would be the correct policy if a new node satisfying more than one `QatDevicePlugin` joins the cluster?

Wouldn't be better to allow only one CR of a certain type per cluster?

Reply:
> Wouldn't be better to allow only one CR of a certain type per cluster?

@rojkov I don't have a good policy in mind. I was thinking several CRs might be handy (especially for QAT) so that I could deploy the plugin using different parameters depending on the node labels. However, now that I think of it, perhaps one CR of a certain type per cluster enforcement is better. Let me digest this.

Reply:
Probably the issue could be solved with adding inter-pod anti-affinity: upon a DS created modify all existing DS to have anti-affinity on the pods created by the new DS (and the new DS itself should create pods with anti-affinity on the pods created by the already existing DS).

Alternatively the QAT plugins could be configured through ENV variables set from the pod labels. Then again only CRD controller would suffice.

Reply:
My thinking, after all, is we keep the controller simple and validate "one CR only".

The failure to deploy multiple device plugin pods of the same type per node comes from the server gRPC socket name. We have `generic` (or `i915`) representing different HW versions. The application can specify node affinity labels to request `generic` resource to be from a certain device.

Reply:
Makes sense. I've added a simple check to the webhooks.

Reply:
thanks,

I'm going to play with the overall lifecycle of the operator next: adding/updating/removing the operator deployment vs. the CRs

Reply:
The book keeping goes wrong if the manager is restarted/updated. Should the check be in the reconcile?

```
default                     qatdeviceplugin-foobar-2wv7v-wd6s7                       0/1     CrashLoopBackOff   1          16s
default                     qatdeviceplugin-sample-vmsbd-jwjl2                       1/1     Running            0          10h
```

Reply:
It can be there too. The problem is that the second CR does exists already in etcd by the time `Reconcile()` is called. In the reconcile we can mark it as failed though and that's it. With webhooks we can prevent registering redundant plugins completely.

My mistake was that I increment the counter upon `create` events only which don't trigger after restarts. Should have used setting the counter upon any update instead of incrementing/decrementing. I keep forgetting that cloud apps should be as stateless as possible :)

Reply:
> Should have used setting the counter upon any update instead of incrementing/decrementing.

OK, if this can be fixed in the webhook we should keep it there.

Reply:
Updated the reconciler to refresh plugin counts on every event. And tested it works now with a restarted operator pod too.

My intuition tells me the current solution is still imperfect though, but i've spent too much time on it already.